File: app/src/main/java/de/tum/in/tumcampusapp/adapters/SimpleStickyListHeadersAdapter.java
package de.tum.in.tumcampusapp.adapters;

import android.content.Context;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.BaseAdapter;
import android.widget.TextView;

import java.util.ArrayList;
import java.util.List;

import de.tum.in.tumcampusapp.R;
import se.emilsjolander.stickylistheaders.StickyListHeadersAdapter;

/**
 * An abstract StickyListHeadersAdapter helps to reduce redundant work for using StickyListHeaders.
 * It implements some method requred by StickyListHeadersAdapter, including getHeaderView getHeaderId,
 * getCount, getItem and getItemId.
 * By extending this class, only getView need to implement.
 * On the other hand this class requires the data model implementing its interface to get header name and id.
 *
 * @param <T> The model of data
 */
abstract public class SimpleStickyListHeadersAdapter<T extends SimpleStickyListHeadersAdapter.SimpleStickyListItem>
        extends BaseAdapter implements StickyListHeadersAdapter {
    List<T> infoList;
    Context context;
    private final List<String> filters;
    final LayoutInflater mInflater;

    SimpleStickyListHeadersAdapter(Context context, List<T> infos) {
        this.context = context;
        this.infoList = infos;
        this.mInflater = LayoutInflater.from(context);

        filters = new ArrayList<>();
        for (T result : infoList) {
            String item = result.getHeaderId();
            if (!filters.contains(item)) {
                filters.add(item);
            }
        }
    }

    // getView mtehod need to be implemented by subclass
    @Override
    abstract public View getView(int position, View convertView, ViewGroup parent);

    public List<T> getInfoList() {
        return infoList;
    }

    public LayoutInflater getInflater() {
        return mInflater;
    }

    /**
     * Genernate header for this item.
     * This methoded can be override if header name need to be modified.
     *
     * @param item the item
     * @return the header for this item
     */
    String genenrateHeaderName(T item) {
        return item.getHeadName();
    }

    @Override
    public View getHeaderView(int position, View convertView, ViewGroup partent) {
        HeaderViewHolder holder;
        View view = convertView;

        if (view == null) {
            holder = new HeaderViewHolder();
            view = mInflater.inflate(R.layout.header, partent, false);
            holder.text = view.findViewById(R.id.lecture_header);
            view.setTag(holder);
        } else {
            holder = (HeaderViewHolder) view.getTag();
        }

        String headerText = genenrateHeaderName(infoList.get(position));
        holder.text.setText(headerText);

        return view;
    }

    @Override
    public long getHeaderId(int i) {
        return filters.indexOf(infoList.get(i)
                                       .getHeaderId());
    }

    @Override
    public int getCount() {
        return infoList.size();
    }

    @Override
    public Object getItem(int position) {
        return infoList.get(position);
    }

    @Override
    public long getItemId(int position) {
        return position;
    }

    // Header view
    private static class HeaderViewHolder {
        TextView text;
    }

    public interface SimpleStickyListItem {
        String getHeadName();

        String getHeaderId();
    }
}

File: app/src/main/java/de/tum/in/tumcampusapp/component/other/generic/adapter/SimpleStickyListHeadersAdapter.kt
package de.tum.`in`.tumcampusapp.component.other.generic.adapter

import android.content.Context
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.BaseAdapter
import android.widget.TextView
import de.tum.`in`.tumcampusapp.R
import se.emilsjolander.stickylistheaders.StickyListHeadersAdapter

/**
 * An abstract StickyListHeadersAdapter helps to reduce redundant work for using StickyListHeaders.
 * It implements some method required by StickyListHeadersAdapter, including getHeaderView getHeaderId,
 * getCount, getItem and getItemId.
 * By extending this class only getView needs to implemented.
 * On the other hand this class requires the data model implementing its interface to get header name and id.
 *
 * @param <T> the data model
</T> */
abstract class SimpleStickyListHeadersAdapter<T : SimpleStickyListHeadersAdapter.SimpleStickyListItem>(
        var context: Context,
        var itemList: MutableList<T>
) : BaseAdapter(), StickyListHeadersAdapter {

    private val filters: MutableList<String>
    val inflater: LayoutInflater = LayoutInflater.from(context)

    init {
        filters = itemList.map { it.getHeaderId() }.distinct().toMutableList()
    }

    // needs to be implemented by subclass
    abstract override fun getView(position: Int, convertView: View?, parent: ViewGroup): View

    override fun getHeaderView(position: Int, convertView: View?, parent: ViewGroup): View {
        val holder: HeaderViewHolder
        val view: View

        if (convertView == null) {
            holder = HeaderViewHolder()
            view = inflater.inflate(R.layout.header, parent, false)
            holder.text = view.findViewById(R.id.lecture_header)
            view.tag = holder
        } else {
            view = convertView
            holder = view.tag as HeaderViewHolder
        }

        val headerText = generateHeaderName(itemList[position])
        holder.text?.text = headerText

        return view
    }

    /**
     * Generate header for this item.
     * This method can be overridden if the header name needs to be modified.
     *
     * @param item the item
     * @return the header for this item
     */
    open fun generateHeaderName(item: T): String = item.getHeadName()

    override fun getHeaderId(i: Int): Long = filters.indexOf(itemList[i].getHeaderId()).toLong()

    override fun getCount(): Int = itemList.size

    override fun getItem(position: Int) = itemList[position]

    override fun getItemId(position: Int): Long = position.toLong()

    // Header view
    private class HeaderViewHolder {
        internal var text: TextView? = null
    }

    interface SimpleStickyListItem {
        fun getHeadName(): String
        fun getHeaderId(): String
    }
}